### ISSUE TITLE
java.lang.IllegalArgumentException when calling ConverterUtil.toPMML

### ISSUE THREAD

**Issue by obones** [jpmml/jpmml-sparkml#22]
Hello,

I'm trying to use pmml export on a very small data sample used with DecisionTreeRegressor and I am getting a `java.lang.IllegalArgumentException`error when calling `ConverterUtil.toPMML`
Here is the code:

    val sourceData = session.read.format("myformat").
      load(DataFileURL)

    val assembler = new VectorAssembler()
      .setInputCols(Array("X1", "X2", "X3"))
      .setOutputCol("features")

    val dt = new DecisionTreeRegressor()
      .setLabelCol("Y")
      .setFeaturesCol("features")
      .setImpurity("variance")
      .setMaxDepth(30)
      .setMaxBins(32)

    val pipeline = new Pipeline().setStages(Array(assembler, dt))

    val model = pipeline.fit(sourceData)

    val pmml = ConverterUtil.toPMML(sourceData.schema, model)

`X1` and `X2` have a `NominalAttribute` attribute while `X3` has the `NumericAttribute`

If I print the `DecisionTreeRegressionModel`, I get this result:

    DecisionTreeRegressionModel (uid=dtr_8e66c6f292fc) of depth 7 with 57 nodes
      If (feature 2 <= 20.0)
       If (feature 1 in {0.0,1.0})
        If (feature 1 in {1.0})
         If (feature 0 in {0.0})
          Predict: 309.38
         Else (feature 0 not in {0.0})
          Predict: 569.6666666666669
        Else (feature 1 not in {1.0})
         If (feature 0 in {0.0})
          Predict: 583.5585714285714
         Else (feature 0 not in {0.0})
          Predict: 591.8775
       Else (feature 1 not in {0.0,1.0})
        If (feature 0 in {1.0})
         Predict: 1882.7800000000002
        Else (feature 0 not in {1.0})
         Predict: 2435.3799999999997
      Else (feature 2 > 20.0)
       If (feature 1 in {0.0})
        If (feature 2 <= 22.0)
         If (feature 0 in {0.0})
          If (feature 2 <= 21.0)
           Predict: 160.80599999999998
          Else (feature 2 > 21.0)
           Predict: 418.02833333333336
         Else (feature 0 not in {0.0})
          If (feature 2 <= 21.0)
           Predict: 636.2533333333334
          Else (feature 2 > 21.0)
           Predict: 273.82000000000005
        Else (feature 2 > 22.0)
         If (feature 2 <= 24.0)
          If (feature 0 in {0.0})
           If (feature 2 <= 23.0)
            Predict: 196.11
           Else (feature 2 > 23.0)
            Predict: 214.44
          Else (feature 0 not in {0.0})
           Predict: 303.5300000000001
         Else (feature 2 > 24.0)
          Predict: 152.13000000000002
       Else (feature 1 not in {0.0})
        If (feature 2 <= 22.0)
         If (feature 1 in {2.0})
          If (feature 2 <= 21.0)
           If (feature 0 in {1.0})
            Predict: 238.91666666666666
           Else (feature 0 not in {1.0})
            Predict: 244.89999999999998
          Else (feature 2 > 21.0)
           Predict: 333.3599999999999
         Else (feature 1 not in {2.0})
          If (feature 2 <= 21.0)
           If (feature 0 in {0.0})
            Predict: 387.8825
           Else (feature 0 not in {0.0})
            Predict: 446.2525
          Else (feature 2 > 21.0)
           If (feature 0 in {1.0})
            Predict: 316.75
           Else (feature 0 not in {1.0})
            Predict: 402.85714285714283
        Else (feature 2 > 22.0)
         If (feature 2 <= 24.0)
          If (feature 0 in {0.0})
           If (feature 2 <= 23.0)
            Predict: 239.59000000000003
           Else (feature 2 > 23.0)
            If (feature 1 in {2.0})
             Predict: 541.51
            Else (feature 1 not in {2.0})
             Predict: 1087.8500000000001
          Else (feature 0 not in {0.0})
           If (feature 2 <= 23.0)
            If (feature 1 in {1.0})
             Predict: 842.3125000000001
            Else (feature 1 not in {1.0})
             Predict: 1059.3700000000003
           Else (feature 2 > 23.0)
            Predict: 384.6300000000001
         Else (feature 2 > 24.0)
          If (feature 1 in {1.0})
           Predict: 350.58000000000004
          Else (feature 1 not in {1.0})
           Predict: 477.96000000000004

What am I missing to be able to get a PMML model?

**Comment by vruusmann** [comment]
> I am getting a java.lang.IllegalArgumentExceptionerror when calling ConverterUtil.toPMML

Would you mind sharing the full stack trace of this IllegalArgumentException?

**Comment by obones** [comment]
Sorry, I forgot to add the call stack where it happens, here it is:

    java.lang.IllegalArgumentException
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeCategoricalSplit(TreeModelUtil.java:287)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeSplit(TreeModelUtil.java:191)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeInternalNode(TreeModelUtil.java:119)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeNode(TreeModelUtil.java:105)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeInternalNode(TreeModelUtil.java:121)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeNode(TreeModelUtil.java:105)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeTreeModel(TreeModelUtil.java:92)
    	at org.jpmml.sparkml.model.TreeModelUtil.encodeDecisionTree(TreeModelUtil.java:64)
    	at org.jpmml.sparkml.model.DecisionTreeRegressionModelConverter.encodeModel(DecisionTreeRegressionModelConverter.java:36)
    	at org.jpmml.sparkml.model.DecisionTreeRegressionModelConverter.encodeModel(DecisionTreeRegressionModelConverter.java:26)
    	at org.jpmml.sparkml.ModelConverter.registerModel(ModelConverter.java:167)
    	at org.jpmml.sparkml.ConverterUtil.toPMML(ConverterUtil.java:81)
    	at com.company.project.DefaultSourceTest$$anonfun$3.apply$mcV$sp(DefaultSourceTest.scala:122)

**Comment by vruusmann** [comment]
> X1 and X2 have a NominalAttribute attribute while X3 has the NumericAttribute

That's a good indication why the "blind" conversion of `VectorUDT` columns is impossible - the vector datatype implies that all vector elements are continuous doubles, whereas in this case there's a mix of categorical and continuous doubles (or even categorical integers and continuous doubles).

> java.lang.IllegalArgumentException
>	at org.jpmml.sparkml.model.TreeModelUtil.encodeCategoricalSplit(TreeModelUtil.java:287)

This exception means that DecisionTreeRegressionModel converter expected to find a categorical feature (because the split type is categorical), but found a continuous feature.

Your categorical features ("feature 0" and "feature 1") appear to be some index values. As a workaround, you should insert a synthetic [`VectorIndexerModel` transformation](https://spark.apache.org/docs/2.1.0/api/java/org/apache/spark/ml/feature/VectorIndexerModel.html) into the first position of the synthetic `PipelineModel` object (ie. in front of the `VectorAssembler` transformation), which explicitly "casts" these two features from continuous value space to categorical value space.

In summary, your feature vector contains heterogeneous data, but it's not reflected in the schema and pipeline model objects that you're passing to the converter utility.

**Comment by obones** [comment]
The thing is that this is not a synthetic pipeline, the code is as above. The `DecisionTreeRegressor` works just fine with that because in the schema, all fields have metadata associated to them which define valid and required attributes.
And clearly that regressor knows how to deal with data that is continuous for some columns and categorical for others, as stated in the documentation.
I am yet to completely understand the `jpmml` code, but I thought it would use the `Attribute.fromStructField` method to get the appropriate `Attribute` instance from where to get the required information.

**Comment by vruusmann** [comment]
> The DecisionTreeRegressor works just fine with that because in the schema, all fields have metadata associated to them which define valid and required attributes.

The metadata that you're speaking of is typically associated with the `Dataset` object, not the `StructType` object that you get when invoking `Dataset#schema()`.

In other words, `DecisionTreeRegressor` gets to collect column information both from `Dataset` and `Dataset#schema()` objects, whereas JPMML-SparkML is limited to the `Dataset#schema()` object. And the latter is incomplete when it comes to figuring out the "definitions" of individual `VectorUDT` elements.

I thought about your situation a bit more, and I think that the best way to make the "cast" from continuous doubles to categorical integers explicit is to prepend `OneHotEncoder` transformation (possibly coupled with helper `StringIndexer` transformation) into your synthetic pipeline model.

**Comment by obones** [comment]
> In other words, DecisionTreeRegressor gets to collect column information both from Dataset and Dataset#schema() objects, whereas JPMML-SparkML is limited to the Dataset#schema() object. And the latter is incomplete when it comes to figuring out the "definitions" of individual VectorUDT elements.

That's where I'm a bit surprised because the `AttributeGroup` attribute that is placed on the `features` field contains in its `attributes` property as many `Attribute` instances as there are columns in the vector and they all define properly the vector elements.

Sure, the `AttributeGroup` may have been created only with `numAttributes` defined, but this is not the case here which explains why the `DecisionTreeRegressor` uses categorical decisions in its resulting tree.

**Comment by vruusmann** [comment]
There is no way to see any `AttributeGroup` or `Attribute` objects if the starting point is a plain `StructType` object as returned by `Dataset#schema()`.

In theory, the `ConverterUtil#toPMML()` utility method could use `Dataset` instead of `StructType` as its first parameter, but in practice you don't want to be dragging your dataset between (J)VMs just to see the number of vector dimensions.

**Comment by vruusmann** [comment]
> the features field contains in its attributes property as many Attribute instances as there are columns in the vector and they all define properly the vector elements.

Perhaps the JPMML-SparkML library should provide another utility method, which re-packages original schema and pipeline model objects into their proper "synthetic" counterparts by taking into consideration all sorts of attribute metadata that is available via the transformed dataframe.

The current solution where every end user has to do the re-packaging work itself is not very scalable.

**Comment by obones** [comment]
> There is no way to see any AttributeGroup or Attribute objects if the starting point is a plain StructType object as returned by Dataset#schema().

Well, to me, there is, as `StructType` has all the right methods to iterate over the `StructField` instances it contains. Once you have the `StructField` for a given column, you can use `Attribute.fromStructField` to get the `Attribute` instance (if any) for that field and work with that.

> Perhaps the JPMML-SparkML library should provide another utility method, which re-packages original schema and pipeline model objects into their proper "synthetic" counterparts by taking into consideration all sorts of attribute metadata that is available via the transformed dataframe.

It would be quite nice to have this, indeed.
